lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,684 @@
+  return ClangASTContext.getConstantArrayType(clangEltTy, size,
+           clang::ArrayType::Normal, 0);
+}
+
+clang::QualType ClangTypeConverter::visitProtocolType(ProtocolType *type) {
+  auto proto = type->getDecl();
+
+  // Single protocol -> id<Proto>
+  if (proto->isObjC()) {
+    auto &clangCtx = ClangASTContext;
+    clang::IdentifierInfo *name = &clangCtx.Idents.get(proto->getName().get());
+    auto *PDecl = clang::ObjCProtocolDecl::Create(
+                    const_cast<clang::ASTContext &>(clangCtx),
+                    clangCtx.getTranslationUnitDecl(), name,
+                    clang::SourceLocation(), clang::SourceLocation(), nullptr);
+
+    // Attach an objc_runtime_name attribute with the Objective-C name to use
+    // for this protocol.
+    SmallString<64> runtimeNameBuffer;
+    PDecl->addAttr(clang::ObjCRuntimeNameAttr::CreateImplicit(
+                     PDecl->getASTContext(),
+                     proto->getObjCRuntimeName(runtimeNameBuffer)));
+
+    auto clangType  = clangCtx.getObjCObjectType(clangCtx.ObjCBuiltinIdTy,
+                                                 &PDecl, 1);
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType ClangTypeConverter::visitMetatypeType(MetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
+
+clang::QualType
+ClangTypeConverter::visitExistentialMetatypeType(ExistentialMetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
+
+clang::QualType ClangTypeConverter::visitClassType(ClassType *type) {
+  auto &clangCtx = ClangASTContext;
+  // produce the clang type INTF * if it is imported ObjC object.
+  auto swiftDecl = type->getDecl();
+  if (swiftDecl->isObjC()) {
+    clang::IdentifierInfo *ForwardClassId =
+      &clangCtx.Idents.get(swiftDecl->getName().get());
+    auto *CDecl = clang::ObjCInterfaceDecl::Create(
+                          clangCtx, clangCtx.getTranslationUnitDecl(),
+                          clang::SourceLocation(), ForwardClassId,
+                          /*typeParamList*/nullptr, /*PrevDecl=*/nullptr,
+                          clang::SourceLocation());
+
+    // Attach an objc_runtime_name attribute with the Objective-C name to use
+    // for this class.
+    SmallString<64> runtimeNameBuffer;
+    CDecl->addAttr(clang::ObjCRuntimeNameAttr::CreateImplicit(
+                     CDecl->getASTContext(),
+                     swiftDecl->getObjCRuntimeName(runtimeNameBuffer)));
+
+    auto clangType  = clangCtx.getObjCInterfaceType(CDecl);
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+  return getClangIdType(clangCtx);
+}
+
+clang::QualType
+ClangTypeConverter::visitBoundGenericClassType(BoundGenericClassType *type) {
+  // Any @objc class type in Swift that shows up in an @objc method maps 1-1 to
+  // "id <SomeProto>"; with clang's encoding ignoring the protocol list.
+  return getClangIdType(ClangASTContext);
```

Comment:
We should probably at least try to preserve type arguments on imported ObjC generic classes.

Reply:
I tried implementing this based on the other things we already have but it didn't work correctly (convert generic args, cast result, use quals() to get protocol decls, use that to create another ObjCObjectType with id as the base).

This doesn't seem super important right now, leaving a TODO instead.